ArrayIndexOutOfBounds when setting enum with very high oid

We've encountered an unusual error on one of our dev environments, where it seems the oid for "enum" is > 2^31 (and therefore negative for signed int), and this is causing pgjdbc-ng to throw an ArrayIndexOutOfBounds exception

```
2021-02-01T04:14:17.932204324Z Caused by: java.lang.ArrayIndexOutOfBoundsException: Index -1846158966 out of bounds for length 4096
2021-02-01T04:14:17.932213915Z 	at java.base/java.lang.invoke.VarHandle$1.apply(Unknown Source)
2021-02-01T04:14:17.932343417Z 	at java.base/java.lang.invoke.VarHandle$1.apply(Unknown Source)
2021-02-01T04:14:17.932367979Z 	at java.base/jdk.internal.util.Preconditions$1.apply(Unknown Source)
2021-02-01T04:14:17.932376899Z 	at java.base/jdk.internal.util.Preconditions$1.apply(Unknown Source)
2021-02-01T04:14:17.932382713Z 	at java.base/jdk.internal.util.Preconditions.outOfBounds(Unknown Source)
2021-02-01T04:14:17.932388616Z 	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Unknown Source)
2021-02-01T04:14:17.932405831Z 	at java.base/jdk.internal.util.Preconditions.checkIndex(Unknown Source)
2021-02-01T04:14:17.932413144Z 	at java.base/java.lang.invoke.VarHandleObjects$Array.compareAndSet(Unknown Source)
2021-02-01T04:14:17.932419435Z 	at java.base/java.lang.invoke.VarHandleGuards.guard_LILL_Z(Unknown Source)
2021-02-01T04:14:17.932544638Z 	at java.base/java.util.concurrent.atomic.AtomicReferenceArray.compareAndSet(Unknown Source)
2021-02-01T04:14:17.932553735Z 	at com.impossibl.postgres.protocol.TypeOid.valueOf(TypeOid.java:43)
2021-02-01T04:14:17.932557924Z 	at com.impossibl.postgres.protocol.v30.MessageDispatchHandler.receiveParameterDescriptions(MessageDispatchHandler.java:496)
2021-02-01T04:14:17.932561965Z 	at com.impossibl.postgres.protocol.v30.MessageDispatchHandler.parseAndDispatch(MessageDispatchHandler.java:311)
2021-02-01T04:14:17.932565935Z 	at com.impossibl.postgres.protocol.v30.MessageDispatchHandler.dispatch(MessageDispatchHandler.java:192)
2021-02-01T04:14:17.932569943Z 	at com.impossibl.postgres.protocol.v30.MessageDispatchHandler.channelRead(MessageDispatchHandler.java:147)
```

This is triggered by executing a query with an enum parameter.

If we take a look at the OID in question

```
select oid, * from pg_type where oid = -1846158966;
-[ RECORD 1 ]--+-----------
oid            | 2448808330
typname        | status_t
typnamespace   | 2448808018
typowner       | 2461479530
typlen         | 4
typbyval       | t
typtype        | e
typcategory    | E
typispreferred | f
typisdefined   | t
typdelim       | ,
typrelid       | 0
typelem        | 0
typarray       | 2448808329
typinput       | enum_in
typoutput      | enum_out
typreceive     | enum_recv
typsend        | enum_send
typmodin       | -
typmodout      | -
typanalyze     | -
typalign       | i
typstorage     | p
typnotnull     | f
typbasetype    | 0
typtypmod      | -1
typndims       | 0
typcollation   | 0
typdefaultbin  |
typdefault     |
typacl         |
```

can see that the oid is high enough to be treated as a negative number. 

It seems to me that this line: https://github.com/impossibl/pgjdbc-ng/blob/415afcf15c2435a6b3f3995b12c0463210151893/driver/src/main/java/com/impossibl/postgres/protocol/v30/MessageDispatchHandler.java#L558 should be using `readUnsignedInt()` rather than `readInt()` - or else `TypeOid` should do a little more checking to ensure that it's not going to exceed the bounds of the array (on the lower side)

I'm not sure how to reproduce this, you'd need to set up a db with enum type having a very high oid. if you can give some tips I'd be happy to do the legwork. 

I will submit a PR, I can test against our dev environment.

--- a/driver/src/main/java/com/impossibl/postgres/protocol/TypeOid.java
+++ b/driver/src/main/java/com/impossibl/postgres/protocol/TypeOid.java
@@ -39,7 +39,7 @@ public class TypeOid implements TypeRef {
 
   public static TypeOid valueOf(int oid) {
     // Craziness is to reduce garbage creation for frequently used types
-    if (oid < fastCachedOids.length()) {
+    if (oid < fastCachedOids.length() && oid >= 0) {
       if (fastCachedOids.compareAndSet(oid, null, INVALID)) {
         // Was null, is now INVALID, set to correct value
         TypeOid toid = new TypeOid(oid);

--- a/driver/src/test/java/com/impossibl/postgres/jdbc/CiTextTest.java
+++ b/driver/src/test/java/com/impossibl/postgres/jdbc/CiTextTest.java
@@ -41,6 +41,7 @@ import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 
+import static org.junit.Assert.assertArrayEquals;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assume.assumeTrue;
@@ -71,6 +72,7 @@ public class CiTextTest {
 
   @Test
   public void testCiTextInSchema() throws SQLException {
+    assumeTrue("testnoexts database is created", TestUtil.isDatabaseCreated("testnoexts"));
 
     String url = "jdbc:pgsql://" + TestUtil.getServer() + ":" + TestUtil.getPort() + "/testnoexts";
 
@@ -102,7 +104,7 @@ public class CiTextTest {
             try (ResultSet rs = stmt.executeQuery("SELECT * FROM citester.test")) {
 
               assertTrue(rs.next());
-              assertEquals(rs.getObject("names", String[].class), new String[] {"test"});
+              assertArrayEquals(rs.getObject("names", String[].class), new String[] {"test"});
 
             }
           }

--- a/driver/src/test/java/com/impossibl/postgres/jdbc/TypeOIDTest.java
+++ b/driver/src/test/java/com/impossibl/postgres/jdbc/TypeOIDTest.java
@@ -1,0 +1,97 @@
+/**
+ * Copyright (c) 2013, impossibl.com
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  * Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *  * Neither the name of impossibl.com nor the names of its contributors may
+ *    be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.impossibl.postgres.jdbc;
+
+import com.impossibl.postgres.api.jdbc.PGConnection;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assume.assumeTrue;
+
+@RunWith(JUnit4.class)
+public class TypeOIDTest {
+
+  private Connection conn;
+
+  @Before
+  public void before() throws Exception {
+    conn = TestUtil.openDB();
+  }
+
+  @After
+  public void after() throws SQLException {
+    TestUtil.closeDB(conn);
+  }
+
+  @Test
+  public void testUnsignedOid() throws SQLException {
+    assumeTrue("Server version is 12 or above", conn.getMetaData().getDatabaseMajorVersion() >= 12);
+
+    try (Connection conn = TestUtil.openDB()) {
+      try (Statement statement = conn.createStatement()) {
+        statement.execute("ALTER SYSTEM SET allow_system_table_mods = 'on'");
+        statement.execute("SELECT pg_reload_conf()");
+      }
+
+      TestUtil.createType(conn, "tester", "value text");
+      try {
+
+        try (Statement statement = conn.createStatement()) {
+          int typeOid;
+          try (ResultSet rs = statement.executeQuery("SELECT oid FROM pg_type WHERE typname = 'tester'")) {
+            assertTrue(rs.next());
+            typeOid = rs.getInt(1);
+          }
+
+          String unsignedTypeOid = Integer.toUnsignedString((int) ((long) Integer.MIN_VALUE + 1024));
+          statement.executeUpdate("UPDATE pg_type SET oid = " + unsignedTypeOid + " WHERE oid = " + typeOid);
+          statement.executeUpdate("DELETE FROM pg_type WHERE typelem = " + typeOid); // clean up orphaned array type
+
+          assertNotNull(conn.unwrap(PGConnection.class).resolveType("tester"));
+        }
+
+      }
+      finally {
+        TestUtil.dropType(conn, "tester");
+      }
+    }
+  }
+
+}

